src/Database/Driver/Sqlserver.php

```diff
@@ -192,17 +193,22 @@ public function enabled(): bool
     public function prepare($query): StatementInterface
     {
         $this->connect();
+
         $options = [PDO::ATTR_CURSOR => PDO::CURSOR_SCROLL];
-        $isObject = $query instanceof Query;
-        /** @psalm-suppress PossiblyInvalidMethodCall */
-        if ($isObject && $query->isBufferedResultsEnabled() === false) {
-            $options = [];
+        if ($query instanceof Query) {
+            if (count($query->getValueBinder()->bindings()) > 2100) {
+                throw new Exception(
```

Comment:
Using core exception feels a bit weird here but we don't have a generic database exception :(

Reply:
I don't know when to use PHP exceptions and when to use Cake exceptions.

Reply:
We often use basic php exceptions when the class fits the situation and we don't have an need to provide a special error page.